helix-term/src/ui/context.rs

```diff
@@ -0,0 +1,510 @@
+        .unwrap_or_else(|| tree.root_node())
+        .byte_range()
+        != tree.root_node().byte_range()
+    {
+        start_node = start_node.expect("parent exists").parent();
+    }
+
+    let context_nodes = doc
+        .language_config()
+        .and_then(|lang| lang.context_query())?;
+
+    let start_index = context_nodes.query.capture_index_for_name("context")?;
+    let end_index = context_nodes
+        .query
+        .capture_index_for_name("context.params")
+        .unwrap_or(start_index);
+
+    let mut cursor = QueryCursor::new();
+    cursor.set_byte_range(start_byte_range.start..context.context_location);
+
+    let query = &context_nodes.query;
+    // Collect the query, for further iteration
+    let query_matches = cursor.matches(
+        query,
+        start_node.unwrap_or_else(|| tree.root_node()),
+        RopeProvider(text),
+    );
+
+    for matched_node in query_matches {
+        // find @context.params nodes
+        let node_byte_range = get_context_paired_range(
+            &matched_node,
+            start_index,
+            end_index,
+            context.topmost_byte,
+            context.context_location,
+        );
+
+        for node in matched_node.nodes_for_capture_index(start_index) {
+            let mut last_node_add = 0;
+            if let Some(last_node) = result.last() {
+                if last_node.line == (node.start_position().row + 1) {
+                    last_node_add += text
+                        .line(text.byte_to_line(context.topmost_byte))
+                        .len_bytes();
+                }
+            }
+
+            if node_in_range(
+                node,
+                context.anchor_line,
+                &node_byte_range.clone(),
+                context.context_location,
+                context.topmost_byte + last_node_add,
+                result.len(),
+            ) {
+                continue;
+            }
+
+            result.push(StickyNode {
+                line: node.start_position().row,
+                visual_line: 0,
+                byte_range: node_byte_range
+                    .as_ref()
+                    .unwrap_or(&(node.start_byte()..node.end_byte()))
+                    .clone(),
+                indicator: None,
+                anchor: view.offset.anchor,
+                has_context_end: node_byte_range.is_some(),
+                view_id: view.id,
+            });
+        }
+    }
+    // result should be filled by now
+    if result.is_empty() {
+        if !cached_nodes.is_empty() {
+            if config.sticky_context.indicator {
+                return Some(add_indicator(&context.viewport, view, cached_nodes));
+            }
+
+            return Some(cached_nodes);
+        }
+
+        return None;
+    }
+
+    let mut res = {
+        cached_nodes.append(&mut result);
+        cached_nodes
+    };
+
+    // Order of commands is important here
+    res.sort_unstable_by(|lhs, rhs| lhs.line.cmp(&rhs.line));
```

Comment:
```suggestion
    res.sort_unstable_by_key(|val| val.line);
```
not that important, but maybe a bit better to read? See https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable_by_key